### PR TITLE
[Feature] Add switch readonly state interface

### PR DESCRIPTION
**Pull request by yzh119**
## Description
Switch between immutable and mutable graph. @BarclayII mentioned about it earlier.
This PR is a part of #369 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
- Add a interface `switch_mode(readonly=True)` in `DGLGraph` class.

### CODE REVIEW

File: python/dgl/graph.py [hunk 1]
```diff
@@ -200,6 +200,37 @@ def __init__(self,
         self._apply_node_func = None
         self._apply_edge_func = None
 
+    def switch_mode(self, readonly=True):
```
Comment (BarclayII) [inline python/dgl/graph.py]: The name `switch_mode` is too generic and doesn't seem to match the docstring.  Either change the name to `switch_readonly`, or the docstring to `Switch mode of the graph` if future toggles are to be included.

File: python/dgl/graph_index.py [hunk 2]
```diff
@@ -138,6 +138,14 @@ def is_readonly(self):
             self._readonly = bool(_CAPI_DGLGraphIsReadonly(self._handle))
         return self._readonly
 
+    def switch_mode(self, readonly=True):
+        """Switch the readonly state of graph index."""
+        n_nodes, multigraph, _, src, dst = self.__getstate__()
+        self.clear_cache()
+        state = (n_nodes, multigraph, readonly, src, dst)
+        self._readonly = readonly
```
Comment (BarclayII) [inline python/dgl/graph_index.py]: `_readonly` should be set in `__setstate__`; change that instead.

**Comment by yzh119** [comment]
@BarclayII I think `mutable_()`/`readonly_()` is a more concise name, what's your idea?

**Comment by VoVAllen** [comment]
How about API like pytorch's `tensor.requires_grad()` and `tensor.requires_grad(False)`, which means `g.readonly()` and `g.readonly(False)`?

**Comment by yzh119** [comment]
@VoVAllen I think `g.readonly_()` is more appropriate considering it's an in-place operation, PyTorch also has a corresponding [requires_grad_](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.requires_grad_) function.

**Comment by BarclayII** [comment]
PyTorch also has `Module.train()` and `Module.eval()` which changes the mode in-place.

I'm fine with either `readonly(bool)` or `switch_mode()`.  `readonly_()` is too PyTorch and I don't like that too much.

**Comment by yzh119** [comment]
I think it sounds more like creating a new graph by calling `g.readonly(bool)`, this is different from `Module.train()` which does not return anything.
However, I have no preference between `readonly(bool)`/`readonly_(bool)`, it's up to you guys.

**Comment by BarclayII** [comment]
I think `readonly` is better.  We don't have the convention that `<func>` creates a new tensor and `<func>_` modifies in-place.

**Comment by yzh119** [comment]
@BarclayII updated.

**Comment by zheng-da** [comment]
i just fixed a bug in the sampler, which fails the CI frequently: https://github.com/dmlc/dgl/pull/401
please rebase to the master branch.

**Pull request merged**